src/ui/SmartSnippet/SmartSnippet.ts

```diff
@@ -0,0 +1,140 @@
+import { exportGlobally } from '../../GlobalExports';
+import { Component } from '../Base/Component';
+import { IComponentBindings } from '../Base/ComponentBindings';
+import { QueryEvents, Initialization, $$ } from '../../Core';
+import { IQuerySuccessEventArgs } from '../../events/QueryEvents';
+import { IQueryAnswerResponse } from '../../rest/QueryAnswerResponse';
+import 'styling/_SmartSnippet';
+import { find } from 'underscore';
+import { IQueryResult } from '../../rest/QueryResult';
+
+const BASE_CLASSNAME = 'coveo-smart-snippet';
+const HAS_ANSWER_CLASSNAME = `${BASE_CLASSNAME}-has-answer`;
+const SHADOW_CLASSNAME = `${BASE_CLASSNAME}-content`;
+const CONTENT_CLASSNAME = `${BASE_CLASSNAME}-content-wrapper`;
+const SOURCE_CLASSNAME = `${BASE_CLASSNAME}-source`;
+const SOURCE_TITLE_CLASSNAME = `${SOURCE_CLASSNAME}-title`;
+const SOURCE_URL_CLASSNAME = `${SOURCE_CLASSNAME}-url`;
+
+export const SmartSnippetClassNames = {
+  HAS_ANSWER_CLASSNAME,
+  SHADOW_CLASSNAME,
+  CONTENT_CLASSNAME,
+  SOURCE_CLASSNAME,
+  SOURCE_TITLE_CLASSNAME,
+  SOURCE_URL_CLASSNAME
+};
+
+export interface ISmartSnippetOptions {}
+
+export class SmartSnippet extends Component {
+  static ID = 'SmartSnippet';
+
+  static options: ISmartSnippetOptions = {};
+
+  static doExport = () => {
+    exportGlobally({
+      SmartSnippet
+    });
+  };
+
+  private shadowContainer: HTMLElement;
+  private sourceContainer: HTMLElement;
+  private snippetContainer: HTMLElement;
+
+  constructor(public element: HTMLElement, public options?: ISmartSnippetOptions, bindings?: IComponentBindings) {
+    super(element, SmartSnippet.ID, bindings);
+    this.bind.onRootElement(QueryEvents.querySuccess, (data: IQuerySuccessEventArgs) => this.handleQuerySuccess(data));
+  }
+
+  private get style() {
+    return $$(this.element)
+      .children()
+      .filter(element => element instanceof HTMLScriptElement && element.type.toLowerCase() === 'text/css')
+      .map(element => element.innerHTML)
+      .join('\n');
+  }
+
+  private set hasAnswer(hasAnswer: boolean) {
+    this.element.classList.toggle(HAS_ANSWER_CLASSNAME, hasAnswer);
+  }
+
+  public createDom() {
+    this.buildShadow();
+    this.buildSourceContainer();
+  }
+
+  private buildShadow() {
+    this.element.appendChild((this.shadowContainer = $$('div', { className: SHADOW_CLASSNAME }).el));
+    const shadow = this.shadowContainer.attachShadow({ mode: 'open' });
+    shadow.appendChild((this.snippetContainer = $$('section', { className: CONTENT_CLASSNAME }).el));
+    shadow.appendChild(this.buildStyle());
```

Comment:
Here since buildStyle can return `undefined`, maybe it would be better to not systematically append it.

Reply:
Woops, you're right.